### ISSUE TITLE
Equals not used when values updated

### ISSUE THREAD

**Issue by SebastianStehle** [immutable-js/immutable-js#1732]
Hi, I was playing around with immutable.js and I like the equals checks. 

What I do not understand is that these checks are not used for updates, e.g. I was playing around with Record and Map:

https://codesandbox.io/s/hidden-sound-h5o2t

**Comment by balagge** [comment]
When you use `.set` on a Record or a Map, the resulting object is normally a "new" object, hence `===` is not suitable for value equality checks.

Some specific scenarios (like `r1_1 === r1_2` in your example) **may happen to** return true, but such behavior should not be assumed/expected.

If you want to compare two values, use `.equals()` (as per your example, `r1_1.equals(r1_2)`. Or, you can also use `immutable.is()`, which is more general (works for any values, not just objects)

https://immutable-js.github.io/immutable-js/docs/#/is

**Comment by SebastianStehle** [comment]
I use `===` for reference equality checks.

In the docs I see the following statement:

> Note: As a performance optimization Immutable.js attempts to return the existing collection when an operation would result in an identical collection, allowing for using === reference equality to determine if something definitely has not changed. This can be extremely useful when used within a memoization function which would prefer to re-run the function if a deeper equality check could potentially be more costly.
https://immutable-js.github.io/immutable-js/

So my assumption was that immutable-js would use the equals helper to check if the same object is added to a collection, e.g. here:

https://github.com/immutable-js/immutable-js/blob/master/src/List.js#L493

**Comment by balagge** [comment]
I think the important part is that it only

> attempts

to return the same collection, but there is no promise to do that. Detecting that an "updating" operation would not actually change anything is sometimes very costly.

The documentation you refer to is a bit vague, but if I understand it correctly it is about a very specific scenario. If you implement a memoizing function, you may do it very thoroughly, i.e. avoid re-running the function for equal inputs. However, if the input is a deep collection, that detection (i.e. calling `is()`, or `.equals()`) may be lengthy and may take as much as re-running the function.

Therefore, what is suggested, is that you may do an "imperfect" version of the memoizing function, which only checks for referential equality (i.e. with `===`). This will NOT handle all "truly equal" cases, but may give an advantage in SOME scenarios -- namely, if the object is actually the same object that has been input previously.

The bottom line is that no immutable operations guarantee that identical (i.e. `===`) objects will be returned. It is only a possibility that may help sometimes.

Using `is()` has little overhead over `===` if the 2 objects are identical. As can be seen from the code, it immediately returns in that case:

https://github.com/immutable-js/immutable-js/blob/751082fd0d392f14f7fa567d5fad6edd8f94bd79/src/is.js#L64-L89

So the bottom line is:
- If you need perfect equality check, only `is()` will give correct results,
- `===` can be used to detect SOME equal scenarios, but if the result is `false` the two objects may still be equal. Howeever, some scenarios can live with this unpredictability and may take advantage of it.

**Comment by balagge** [comment]
as for the code line you refer to, that is quite deep within immutable.js, I do not know what is going on there. It may be that it uses the above mentioned "imperfect" equality check to gain advantage in some scenarios. It is also possible that the specific case involves immutable primitives (numbers or strings) only, so `===` can be used safely.

**Comment by balagge** [comment]
as for your original example sandbox: the returned "updated" collections are identical for numeric values and not identical where the value is a `Set`. Probably a `===` check is used internally to detect "no change" scenarios, because that has not much cost. since `3 === 3`, this scenario is easily captured and the original object is returned. Deeper check is not done, because that would degrade performance in a vast number of cases (i.e. where the new, "updated" value is different).

If you are curious, you could try updating with the exact same object. (i.e. save an instance of `Set("a")` and update with it twice). I **suspect** then you will get back the same object, but I do not have time to try it.

**Comment by SebastianStehle** [comment]
Yes, this is what I wanted to say. I don't understand why "imperfect" equality checks are done to detect no-change scenarios in the collection and why the `is` method is not used there. For me it is unexpected because of these reasons:

1. I assume that in general reading immutable structures happens much more often then updates. And calling a memoized function when not needed can be costly. So it is important that an immutable data structure only creates a new instance when there is a change.
2. When you have deeply nested structures a change in a leaf bubbles up and the additional updates are probably much more expensive than the "perfect" equal check in the leaf.
3. The equal check with `is` is usually very fast, because in most cases it would be just a reference equality for objects (`===`)
4. In Java, C# and probably much more other languages the implementer of the equals method has control, not the collection or data structure. So if your "perfect" equals check is too slow you can always fall back to a "imperfect" solution. But in immutable.js I have no option to control the equal checks.

**Comment by bdurrer** [comment]
Immutable _will_ know when any value changed, even deep within, because then it has to create a new immutable object
It compares by reference, not by value!

```
const r2_1 = A({ b: Set("a") });
const r2_2 = r2_1.set("b", Set("a"));
```

That the check on record returns false is no suprise, as Set("a") creates a new set, and Set("a") !== Set("a")

**Comment by SebastianStehle** [comment]
I understand the "what", but not the "why"...

**Comment by bdurrer** [comment]
Let's take a Java example: primitives and strings are compared by value, right? 

```
String a = "hello";
String b = "hello";
(a == b) is true
```

But objects are compared by reference:
```
House a = new House();
House b = new House();
a == b is false
```
Set("a") creates a new instance of Set. So if you do it twice, you get two different Sets. 
You might implement an equals method, but even then, the == tests the `reference`

Immutable would have to deeply check values at every level of nested collections (!) at the cost of performance, implement equals functions and deal with custom code that might or might not implement equals. It might have to deal with objects that have a wrong implementation of equals ... only for the purpose of returning the same, unchanged collection so that you can use a reference check (===). That doesn't sound like a good idea, doesn't it?

Immutable uses [trie](https://en.wikipedia.org/wiki/Trie) to manage the data. Even updates in nested maps, which result in the creation of a new map at every level, are a relatively fast operation:

const x = Map({ a: Map({ b: Map({ c: "hello" }) })  })
x.setIn(['a', 'b', 'c'], "world"); // 3 new collections!

If each Map's operation had to call `equals` and do deep checks, that would likely be either complicated or bad for performance.

If you think that you run into problems because you have two instances with the same value, you likely did something wrong further up in your code OR you could use Immutable.is before updating an collection.

PS: The following example would return true:
```
const x = Set("a")
const r2_1 = A({ b: x });
const r2_2 = r2_1.set("b", x);
```

**Comment by SebastianStehle** [comment]
If I implement equals myself then it is my responsibility to provide a solution that is fast and good enough for my purposes.

The same is true in all programming languages. e.g. when I use a hash map and I always return a random value when I override the hash code function then I will have a problem. And you have all opportunities, you can use reference equality in your equals method, you can implement id equality (for entities) or value equality for structures like vectors or other value objects.

About performance: This really matters on your use case. Lets say you have a list of 10.000 vectors and your vector is a custom class with equal function, as also described in the docs:

> When data is passed from above rather than being subscribed to, and you're only interested in doing work when something has changed, you can use equality.

When you make an update you have to compare two values. When you want to check if the list has been mutated and you cannot trust the reference equality then you have to compare two lists with equals and you have to compare 10.000 vectors in the worst case.

Btw: I am not using immutable-js anymore, it is just an interesting discussion.

**Comment by SebastianStehle** [comment]
If I implement equals myself then it is my responsibility to provide a solution that is fast and good enough for my purposes.

The same is true in all programming languages. e.g. when I use a hash map and I always return a random value when I override the hash code function then I will have a problem. And you have all opportunities, you can use reference equality in your equals method, you can implement id equality (for entities) or value equality for structures like vectors or other value objects.

About performance: This really matters on your use case. Lets say you have a list of 10.000 vectors and your vector is a custom class with equal function, as also described in the docs:

> When data is passed from above rather than being subscribed to, and you're only interested in doing work when something has changed, you can use equality.

When you make an update you have to compare two values. When you want to check if the list has been mutated and you cannot trust the reference equality then you have to compare two lists with equals and you have to compare 10.000 vectors in the worst case.

Btw: I am not using immutable-js anymore, it is just an interesting discussion.

**Comment by bdurrer** [comment]
It's not about the implementation of equals having to be fast, it's the fact that doing a reference check is so much faster than performing a deep-comparison. The current, reference-based comparison works regardless of what the user does, which is not guaranteed with equals.

When you are performing a mutation, you do not need to check the full collection for changes - because you know exactly what part you touch! all you have to do is check the existing value inside the collection against what you are gonna change.

Forced use of an deep `equals` check would likely make us drop immutable from the project, as modifying the list of 50'000 Records (with sub-collections) would freeze the application.

If it depends on the usecase, it should not be part of the framework.

Anyway, it's not that hard for you to do it before modifying:

const x = Set("a")
const y = Set("a")
const rec1 = A({ b: x });

const willChange = !Immutable.is(rec1.get('b'), y)
const rec2 = rec.set("b", x);

**Comment by SebastianStehle** [comment]
I don't say that you should always make a deep equals of the old and new collection after a collection is modified, I am saying that the objects to update should be compared with equals (if it is implemented).

e.g.

```
// Lets implement a simple readonly array.
function replace(index, newItem) {
  if (!equals(this.items[index], newItem)) {
    const newItems = [...this.items];
    newItems[index] = newItem;

    return new ReadOnlyArray(newItems);
  }

 return this;
}
``
or

// Lets implement a simple readonly array.
function replace(index, newItem) {
  const currentItem = this.items[index];
  if (currentItem !== newItem && !equals(currentItem, newItem)) { // Performance optimization.
    const newItems = [...this.items];
    newItems[index] = newItem;

    return new ReadOnlyArray(newItems);
  }

 return this;
}

instead of
```
// Lets implement a simple readonly array.
function replace(index, newItem) {
  if (this.items[index] !== newItem) {
    const newItems = [...this.items];
    newItems[index] = newItem;

    return new ReadOnlyArray(newItems);
  }

 return this;
}
```

If I do not implement equals, it is exactly the same behavior.

You could even inject an comparer, e.g. in .NET you have `IEqualityComparer`, so that you can decide how nested objects or collections should be compared.

The current strategy is inconsistent in my opinion and impossible to adapt to my needs. Because in my use case I render stuff using canvas or svg and this is much more expensive than equal comparisons.

In your use case it would also create no performance problems, because you are not modifying 50.000 nested collections simultaneously. If you do so, you can also create a new collection. You only have to make the equals check if the reference has not changed in an update and this is in most of the time not the case.

**Comment by bdurrer** [comment]
What you propose is partially already there. The `updater` function of [update()](https://immutable-js.github.io/immutable-js/docs/#/Map/update) allows to check the value and return the original one, if you think it did not change. This would result in the collection not being changed.

> In your use case it would also create no performance problems, because you are not modifying 50.000 nested collections simultaneously. If you do so, you can also create a new collection

I assume you know that every modification in immutable results in "a new collection"? Replacing a sub-collection with an almost-identical sub-collection would require the root collection to do a deep-equal, so yes it certainly would run into full checks quite often in my case.

> The current strategy is inconsistent in my opinion and impossible to adapt to my needs. Because in my use case I render stuff using canvas or svg and this is much more expensive than equal comparisons.

I don't understand the reasoning behind this statement. 
You seem to arguing for equality checks because you data is _similar_ but not _equal_. What keeps you from doing an equality check before calling set (as shown above)? There is even an `equals` on collections and the `is` function you could use. Why do you create a new collection with Set(...) that has the same content in the first place? 


Anyway my time is used up, I'll let this stand as it is and maybe someone adds an optional-equality-checker feature for `set` operations :)
